feat: css editor

Shobhit-Nagpal opened the pull request:
Hey, opening a PR for implementing the CSS editor for custom styles building on top of #1514. Summary of changes:

- Compared between Prismjs and Shiki for syntax highlighting (Prismjs was giving a lesser bundle size)
- Shifted the CSS editor inside a modal which is being lazy loaded
- Implemented endpoints for posting and restoring css
- Added a prebuild script to write bundled css to the override.css file (do let me know if any changes are required here)
- Enabling the Edit CSS button when the override styles value is true

Here's a video to show the working:

https://github.com/user-attachments/assets/e9852081-7a3a-4fb8-853a-d5ba34b04693

Do let me know if any changes are required

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditor.tsx:
```diff
@@ -0,0 +1,43 @@
+import { useEffect, useState } from 'react';
+import Editor from 'react-simple-code-editor';
+import Prism from 'prismjs';
```
question: could we minimise the impact on bundle size by using the import 
`import Prism from 'prismjs/components/prism-core';`

Shobhit-Nagpal replied:
optimized the bundle size. should be good on this end. working on other changes mentioned

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditor.tsx:
```diff
@@ -0,0 +1,43 @@
+import 'prismjs/components/prism-json';
+
+import 'prismjs/components/prism-css';
+import 'prismjs/themes/prism-funky.min.css';
```
Can we use the tomorrow night?

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditorModal.tsx:
```diff
@@ -0,0 +1,105 @@
+        <ModalHeader>Edit CSS</ModalHeader>
+        <ModalCloseButton />
+        <ModalBody>
+          <CodeEditor onChange={(updatedCss: string) => setCSS(updatedCss)} initialValue={css} language='css' />
```
this structure would effectively cause the entire modal to re-render on every change of the code editor.

Can we isolate the `onChange` inside the `<CodeEditor>` component?

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditorModal.tsx:
```diff
@@ -0,0 +1,105 @@
+import { useEffect, useState } from 'react';
+import {
+  Button,
+  Flex,
+  HStack,
```
cleanup: we dont use layout components from Chakra and have our own. You can search other interfaces in settings to see how things are composed or I can change if you prefer

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditorModal.tsx:
```diff
@@ -0,0 +1,105 @@
+  }, []);
+
+  return (
+    <Modal isOpen={isOpen} onClose={onClose} variant='ontime' isCentered>
```
question: what does the isCentered do? we dont have it in any other modals and they are all in the center

Shobhit-Nagpal replied:
in case the editor has no content, the modal is towards the top. the isCentered here is to keep the modal in center even if there's no content. attaching a preview for what happens when we remove this prop and the css is empty

![image](https://github.com/user-attachments/assets/d4c03709-c468-4d80-a25d-5d715481ad9d)

cpvalente replied:
I see, thank you for clarifying.

I think what we need to do is to give the editor a fixed size and remove the is centered prop

Shobhit-Nagpal replied:
yes, this seemed to work. added a min height for the editor as you had suggested, works well now

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/ViewSettingsForm.tsx:
```diff
@@ -14,11 +14,14 @@ import { preventEscape } from '../../../../common/utils/keyEvent';
 import { isOntimeCloud } from '../../../../externals';
 import * as Panel from '../../panel-utils/PanelUtils';
 
+const CodeEditorModal = lazy(() => import('./StyleEditorModal'));
```
question: what do you think about lazy loading the `<CodeEditor>` instead?
I wonder if we can get it to load while we fetch the CSS?

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/ViewSettingsForm.tsx:
```diff
@@ -115,6 +119,17 @@ export default function ViewSettingsForm() {
                   <Switch variant='ontime' size='lg' isChecked={value} onChange={onChange} ref={ref} />
                 )}
               />
+              <Panel.InlineElements>
```
question: why do we need the `<InlineElements>` component?

Shobhit-Nagpal replied:
without it, the button is taking full width. attaching a preview:

![image](https://github.com/user-attachments/assets/04dd7b32-2319-40ce-8d57-3a3ac2dd56c3)

do you feel it's better to manipulate the width of the button through the button's props itself?

cpvalente replied:
yes, I think this is a CSS problem, not a layout one.
I dont think we should solve it by adding more elements to the DOM

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditor.tsx:
```diff
@@ -0,0 +1,43 @@
+  onChange: (newValue: string) => void;
+}
+
+const CodeEditor = (props: CodeEditorProps) => {
```
questions: 
- we usually use the function keyword and export default inline, is there a reason not to do this here?
- is this component a good candidate for memoisation? it sounds like it can be slow to render

cpvalente commented on apps/server/src/api-data/db/db.controller.ts:
```diff
@@ -311,3 +315,54 @@ export async function deleteProjectFile(req: Request, res: Response<MessageRespo
     res.status(500).send({ message });
   }
 }
+
```
it feels a bit strange to me to have this code here, but I dont know where to put it. What do you think of the following?

- the controllers shouldnt really be writing to disk, we would prefer having a very thin service to do the writing part
- how about creating an `asset` directory with `asset.router` + `asset.controller` + `assert.service` for this?

Shobhit-Nagpal replied:
i like this idea, i'll make the changes accordingly

cpvalente reviewed (commented):
Good work, I left some small comments and suggestions for cleanup that you can consider

Shobhit-Nagpal commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditor.module.scss:
```diff
@@ -0,0 +1,10 @@
+.code {
+  font-family: monospace;
+  font-size: 1rem;
+}
```
not in prism but the code editor seems to handle it. made the changes accordingly

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditorModal.tsx:
```diff
@@ -0,0 +1,109 @@
+              </Button>
+            </div>
+            <div>
+              <div className={style.buttonRow}>
```
we have the Panel.InlineElements which we should use instead

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/ViewSettingsForm.tsx:
```diff
@@ -115,6 +119,16 @@ export default function ViewSettingsForm() {
+                size='sm'
+                marginTop='1rem'
+                isDisabled={!data.overrideStyles}
+                width={24}
```
question: 
I wonder if we can improve the CSS here
- layout items like margin-top should come from the parents (ideally), is that possible here?
- the width 24 seems strange, we dont want to hard code a width, we just want to stop if from being 100%. So the correct value would be like `width: fit-contents`

Shobhit-Nagpal replied:
> layout items like margin-top should come from the parents (ideally), is that possible here?

looked into this, one way this is possible is if we add put the button under a `div` and add margin-top to that, but that's rendering an extra node only to be able to put the margin.

Something similar is done in the `<GenerateLinkForm>` under the Network Panel with more styling [here](https://github.com/cpvalente/ontime/blob/73afbacdd76c2259c0b4caeebe852730a6ffc9ec/apps/client/src/features/app-settings/panel/network-panel/GenerateLinkForm.tsx#L128)

One option is to put a `row-gap` under the `.listItem` styling being applied on the `<Panel.ListItem>` component [here](https://github.com/cpvalente/ontime/blob/73afbacdd76c2259c0b4caeebe852730a6ffc9ec/apps/client/src/features/app-settings/panel-utils/PanelUtils.tsx#L90). This eliminates the need for adding a margin-top on the new rows that come in under `<Panel.ListItem>`

cpvalente commented on apps/server/scripts/bundleCss.ts:
```diff
@@ -0,0 +1,25 @@
+    await writeFile(cssFile, defaultCss, { encoding: 'utf8' });
+  } catch (error) {
+    console.error('Failed writing to CSS file: ', error);
+    process.exit(1);
```
cleanup: I dont think we need this, the program will terminate after this line anyway and there is nothing catching this directive

cpvalente commented on apps/server/src/api-data/assets/assets.router.ts:
```diff
@@ -0,0 +1,8 @@
+
+export const router = express.Router();
+
+router.get('/css', getCssOverride);
```
nice!

cpvalente commented on apps/server/src/api-data/assets/assets.service.ts:
```diff
@@ -0,0 +1,31 @@
+ */
+export async function readCssFile(): Promise<string> {
+  const path = publicFiles.cssOverride;
+  if (!existsSync(path)) {
+    throw new Error('File not found');
+  }
```
question: if the path does not exist (ie: the user deleted the file mid session), should we write the contents of the bundled css to a new file?

cpvalente commented on apps/server/src/api-data/assets/assets.service.ts:
```diff
@@ -0,0 +1,31 @@
+ */
+export async function writeCssFile(css: string) {
+  const path = publicFiles.cssOverride;
+  if (!existsSync(path)) {
+    throw new Error('File not found');
+  }
```
question: if the CSS file does not exist, should we make one to write the contents to?

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditor.tsx:
```diff
@@ -0,0 +1,57 @@
+        style={{
+          fontFamily: 'monospace',
+          fontSize: 12,
+          minHeight: 500,
```
question: does the minHeight mean that the contents can grow? we likely do not want that. we want the editor to have a fixed size with content scrolling if necessary

Shobhit-Nagpal replied:
the minHeight is to give a minimum height in case there's no css so that the modal still retains its structure.

for the content growing, it's not growing with content. if there's more content than the height, then scrolling kicks in. attaching a preview below:


https://github.com/user-attachments/assets/5c0cbb09-8f83-4775-af23-ca8d77bddef8

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditor.tsx:
```diff
@@ -0,0 +1,57 @@
+    setCode(initialValue);
+  }, [initialValue]);
+
+  useEffect(() => {
```
cleanup: can you please write a comment to what these useEffects achieve? like

`sync editor contents to external source`
and
`add contents to editor on mount`

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditorModal.tsx:
```diff
@@ -0,0 +1,109 @@
+  const [saveLoading, setSaveLoading] = useState(false);
+  const [resetLoading, setResetLoading] = useState(false);
+
+  const cssRef = useRef<string>(css);
```
question: I wonder if the CSS ref is misplaced

It feels strange for the parent to control the contents of the child this way. 
The UI should be a function of state.
When the reset function is called, the contents change, this state should trigger a re-render which updates the Editor without us explicitely doing so

cpvalente commented on apps/client/src/features/app-settings/panel/general-panel/StyleEditorModal.tsx:
```diff
@@ -0,0 +1,109 @@
+
+  const cssRef = useRef<string>(css);
+
+  const handleReset = async () => {
```
nitpick: should this be called `handleRestore`? resetting and restoring provide different expectations as to what will happen

cpvalente reviewed (commented):
Great work, I didnt get to test it but had a look at the code and I think we can do some minor improvements

I will try to find some time to test the usability of it, but assume everything is OK

Shobhit-Nagpal commented:
hey, made changes according to the previous review:

- removed `process.exit` in catch block
- changed `handleReset` to `handleRestore`
- added comments on useEffect
- added modal button row under `<Panel.InlineElements>`
- handled modal render and the ref misplacement by using `useImperativeHandle` to get the changed css from the `StyleEditor` (UI should be a function of state now and not dependent on the ref for any changes on the editor)
- handled file not existing by creating the file with the bundledCss instead of throwing an error
- added disabling/enabling of save button according to the contents of the editor

let me know if any further changes are required
